### DISCUSSION TITLE
Does Megatron support tracing computation graphs with torch.fx?

### DISCUSSION THREAD

**Discussion by fy-j** [NVIDIA/Megatron-LM#1323] [Q&A]
I am trying to trace a computation graph in Megatron using torch.fx. However, I encountered the following error:
```shell
[rank1]: Traceback (most recent call last):
[rank1]:   File "/kimchou/GeeS***/python/gees***/adapters/pytorch/getTorchGraph.py", line 58, in getTorchGraph
[rank1]:     traced = symbolic_trace(model) # use transformers.utils.fx to trace the model
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/torch/fx/_symbolic_trace.py", line 1193, in symbolic_trace
[rank1]:     graph = tracer.trace(root, concrete_args)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/eval_frame.py", line 437, in _fn
[rank1]:     return fn(*args, **kwargs)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/torch/_dynamo/external_utils.py", line 36, in inner
[rank1]:     return fn(*args, **kwargs)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/torch/fx/_symbolic_trace.py", line 793, in trace
[rank1]:     (self.create_arg(fn(*args)),),
[rank1]:   File "/kimchou/Megatron-LM-test/megatron/core/models/gpt/gpt_model.py", line 343, in forward
[rank1]:     **(extra_block_kwargs or {}),
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/torch/fx/proxy.py", line 443, in __bool__
[rank1]:     return self.tracer.to_bool(self)
[rank1]:   File "/usr/local/lib/python3.10/dist-packages/torch/fx/proxy.py", line 303, in to_bool
[rank1]:     raise TraceError('symbolically traced variables cannot be used as inputs to control flow')
```
How can I use torch.fx to trace the model graph

**Comment by Advait251206** [comment]
The error you're seeing is a common limitation of **`torch.fx.symbolic_trace`**, rather than something specific to Megatron-LM.

The key part of the traceback is:

```text
TraceError: symbolically traced variables cannot be used as inputs to control flow
```

This occurs because `torch.fx` represents tensors as `Proxy` objects during tracing. Whenever the model executes Python control flow based on one of these proxies (e.g., `if`, `and`, `or`, `bool()`, `len()`), the tracer cannot determine which branch to take and raises a `TraceError`.

In your case, the traceback points to:

```python
**(extra_block_kwargs or {}),
```

During tracing, `extra_block_kwargs` becomes a `Proxy`, and Python attempts to evaluate its truth value via the `or` operator, which internally calls `__bool__()`. Since `Proxy.__bool__()` is intentionally unsupported, tracing fails.

---

## Is Megatron officially compatible with `torch.fx`?

At the moment, **Megatron-LM is not designed to be fully traceable using `torch.fx.symbolic_trace`**.

There are several reasons for this:

- Dynamic Python control flow in model code.
- Distributed communication primitives (TP, PP, CP).
- Custom CUDA/Transformer Engine operators.
- Runtime-dependent execution paths.
- Backend dispatch logic.

All of these make symbolic tracing difficult with the current FX tracer.

---

## Possible workarounds

### 1. Use `concrete_args`

If the problematic arguments are effectively constants during tracing, you can provide them as concrete values:

```python
from torch.fx import symbolic_trace

traced = symbolic_trace(
    model,
    concrete_args={
        "extra_block_kwargs": None,
    },
)
```

This works if those arguments are not expected to change dynamically.

---

### 2. Refactor Python control flow

FX can only trace code whose control flow does **not** depend on symbolic tensors.

For example, instead of

```python
kwargs = extra_block_kwargs or {}
```

use something like

```python
if extra_block_kwargs is None:
    kwargs = {}
else:
    kwargs = extra_block_kwargs
```

provided `extra_block_kwargs` is a normal Python object and not a traced tensor.

Similarly, avoid constructs such as

```python
if tensor:
```

or

```python
if proxy:
```

since these cannot be symbolically evaluated.

---

### 3. Use `torch.export` (PyTorch 2.x)

If your goal is graph extraction rather than FX transformations, `torch.export` is generally a better choice for modern PyTorch models.

Compared to `torch.fx.symbolic_trace`, `torch.export` supports a much wider range of dynamic behaviors and is becoming the recommended graph capture mechanism in PyTorch 2.x.

---

### 4. Use `torch.compile` / TorchDynamo

If your objective is optimization rather than graph visualization, `torch.compile` (TorchDynamo + AOTAutograd) is usually a better fit.

TorchDynamo captures graphs by intercepting Python bytecode instead of relying purely on symbolic execution, making it compatible with many models that cannot be traced using vanilla FX.

---

### 5. Trace individual submodules

If you're primarily interested in the transformer computation itself, you may have better success tracing smaller components (e.g., an MLP block or attention block) instead of the entire distributed `GPTModel`.

Many of the tracing issues arise from the top-level model orchestration rather than the individual neural network layers.

---

## If your goal is graph visualization

If you're simply trying to inspect the computation graph, alternatives such as:

- `torch.export`
- `torch.compile` (with graph dumping)
- ONNX export (where applicable)
- `torchviz`
- Nsight Systems / Nsight Compute (for execution profiling)

may provide a more reliable representation than `torch.fx.symbolic_trace` for Megatron models.

---

## Conclusion

The error isn't caused by an incorrect usage of `torch.fx`; it's a consequence of how `torch.fx.symbolic_trace` works. Megatron-LM contains dynamic Python control flow and distributed execution logic that are not fully compatible with FX's symbolic tracer. If you only need to trace portions of the model, supplying `concrete_args` or tracing individual submodules may be sufficient. Otherwise, for full-model graph capture, I'd recommend looking at **`torch.export`** or **`torch.compile`**, which are better suited to modern, highly dynamic models like Megatron-LM.

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206
